LiteCore/Database/DatabaseImpl.cc

```diff
@@ -642,30 +659,46 @@ namespace litecore {
 
         // Instantiate it, creating the KeyStore on-disk if necessary:
         KeyStore &store = _dataFile->getKeyStore(keyStoreName);
-        auto collection = make_retained<CollectionImpl>(this, spec, store);
-        // Update its state & add it to _collections:
-        auto collectionPtr = collection.get();
-        _collections.insert({CollectionSpec(collection->getSpec()),
-                             move(collection)});
-        if (isInTransaction())
-            collectionPtr->transactionBegan();
+        if(!collectionPtr) {
+            // This will be entered if the collection never existed or was removed, but
+            // skipped if there is already a collection that was previously invalidated 
+            // (the latter case already reset valid to true near the top of this method)
+            auto collection = make_retained<CollectionImpl>(this, spec, store);
+            // Update its state & add it to _collections:
+            collectionPtr = collection.get();
+            _collections.insert({CollectionSpec(collection->getSpec()),
+                                 move(collection)});
+
+            if (isInTransaction())
+                collectionPtr->transactionBegan();
+        }
+
+        
+
+        collectionPtr->setValid(true);
```

Comment:
As I noted, we only  setValid(false) for other db instance. We deleted underlying sqlite table. Now, we reuse the collection object on logically completely different collection sharing the collection name.

Reply:
I don't see the issue with that.  If you recreate the table, then everything is ok again.  We reset to valid, and the table is there.

Reply:
Perhaps we have different mental model regarding delete and recreate the collection. In my model, deleted and recreated collections have nothing to do with each other except they share the same name. This is something confuses me.

Reply:
Perhaps we have different mental model regarding delete and recreate the collection. In my model, deleted and recreated collections have nothing to do with each other except they share the same name. This is something confuses me.

Reply:
A collection is nothing more than a logical wrapper around a table, like a pointer.  So if you restore the underlying data then in my mind you've restored the contents of the wrapper, then there is no significant difference between that and creating a brand new object just to point to the same table name again.  However, I suppose that if the old one were removed at the time of the new one's creation that would not present any threading issues so that might be a bit more clean.  That is to say, if you ask a given database for a collection, and it finds one that is invalid, it will try to retrieve it again from disk and if it finds it then instead of resurrecting the old one, it removes it and creates a new one (if the consumer obeys the threading model then all of this will happen on one thread, so it's ok, as opposed to removing it from the external notification which is running on another thread probably).  Is that in line with what you are thinking?

Reply:
I think collection observers are associated with the collection object. Now, if I delete the collection, what does it mean to the observers? The observers now don't have the object to observe. At this time, they exist to observe what ever in the future a collection with the same name is created. Does the observer observe this collection, or any collection of that name? If any collection by that name, then is it problem to erase the collection from _collections of the active db instance?